### ISSUE TITLE
Adding local linear regression objective functions

### ISSUE THREAD

**Issue by peterfoley** [dmlc/xgboost#1332]
I've got working code in a (currently private) fork with objective functions to fit a local regression to each observation.

The inputs for this approach are:
- **X[N,k]** the usual design matrix
- **labels[N]** the usual vector of continuous/binary labels
- **locX[N,m]** a new matrix of local regression parameters

And the resuling model outputs are:
- **Beta[N,m+1]** local regression parameters. These are the parameters that are fitted, grouped, boosted by gbtree (reusing `num_output_group` param from `multi:softprob`). 

And the predicted labels are:
`preds[i]  = Beta[i,0] + sum(locX[i,] * Beta[i,1:m])`
and then those predictions, and their derivatives are passed to the other loss functions already implemented like  `reg:linear`, `binary:logistic`, etc. and the chain rule applied to calculate the actual derivatives reported back to gbtree.

In my current code, `locX` is tacked onto `labels` in R (locX is transposed and rbind'ed to labels so that observations' data are kept nearby in memory) before passing it over to xgboost, and some of the data structure checks in the R package code had to be tweaked to let that happen. That worked for my proof of concept but is a very messy solution and will need to be redesigned for the final implementation.

The final implementation will also need to ensure that the data structures and parameters are consistent in the vector regression code and this new local linear regression code.

Should I kick off a new pull request and start [this process](http://xgboost.readthedocs.io/en/latest/how_to/contribute.html) or should I break apart the parameter passing changes from the local linear regression? My changes are only currently tested in the R API, so others will need to help implement parameter changes in the APIs of other languages. The meat of the objective function is implemented in C++, though, so I'd expect only parameter changes to be language-specific.

**Comment by pommedeterresautee** [comment]
Old offer of PR, no answer. Closing.
